src/utils/tests/AlgorithmTest.cpp

```diff
@@ -131,6 +131,54 @@ BOOST_AUTO_TEST_CASE(EmptyRange)
 
 BOOST_AUTO_TEST_SUITE_END() // Range
 
+BOOST_AUTO_TEST_SUITE(ReorderArray)
+
+BOOST_AUTO_TEST_CASE(OneElement)
+{
+  PRECICE_TEST(1_rank);
+  std::array<int, 1> input{1};
+  std::array<int, 1> order{0};
+  std::array<int, 1> expected{1};
+
+  auto reordered = utils::reorder_array(order, input);
+  BOOST_TEST(reordered == expected);
+}
+
+BOOST_AUTO_TEST_CASE(AlreadySorted)
+{
+  PRECICE_TEST(1_rank);
+  std::array<int, 3> input{3,4,5};
+  std::array<int, 3> order{0,1,2};
+  std::array<int, 3> expected{3,4,5};
+
+  auto reordered = utils::reorder_array(order, input);
+  BOOST_TEST(reordered == expected);
+}
+
+BOOST_AUTO_TEST_CASE(Reverse)
+{
+  PRECICE_TEST(1_rank);
+  std::array<int, 3> input{3,4,5};
+  std::array<int, 3> order{2,1,0};
+  std::array<int, 3> expected{5,4,3};
+
+  auto reordered = utils::reorder_array(order, input);
+  BOOST_TEST(reordered == expected);
+}
+
+BOOST_AUTO_TEST_CASE(Scramble)
+{
+  PRECICE_TEST(1_rank);
+  std::array<int, 3> input{3,4,5};
+  std::array<int, 3> order{2,0,1};
+  std::array<int, 3> expected{5,3,4};
+
+  auto reordered = utils::reorder_array(order, input);
+  BOOST_TEST(reordered == expected);
+}
+
+BOOST_AUTO_TEST_SUITE_END() // ReorderArray
```

Comment:
Should we also add a test where `T` and `Index` don't have the same typ? Or I am misunderstanding sth here?